Improve formatting of timespan

iSazonov opened the pull request:
Close #2595 

After this PR:
```powershell
PS C:\> [timespan]"1:12:23:56"

1.12:23:56

PS C:\> Get-Uptime

8.04:44:37
```

iSazonov committed:
Improve formatting of timespan

RichardSiddaway commented:
That's not an improvement.

lzybkr commented:
I agree with @RichardSiddaway - it's not an improvement. I'd really like to see my human friendly formatting, but implemented so that it's more friendly to localization.

iSazonov commented:
@lzybkr I like human friendly display too. The problem is that the solution is a _function_. In the worst case, we need to write a separate function for each language to do localization in full accordance with the rules of the language. It is a matter for the individual project [Humanizer ].(https://github.com/Humanizr/Humanizer) Can we import this project in Powershell? If not, then we need a compromise. What would it be?

This may be some kind of international format (ISO 8601) or .net timespan format (the PR use it).

Finally, we can just leave it 'as is' now.

So we need a conclusion (committee?) on this.

lzybkr commented:
I'm comfortable speaking for the @PowerShell/powershell-committee here (I am a member) - this doesn't really fall under the larger design issues the committee is focused on.

The committee did recently take a very quick look at Humanizer. I'm not sure we'll include it, but it is interesting.  That said, I think my human friendly TimeSpan is nicer than what Humanizer does, and I think it's not that hard to create in a localization friendly manner.

iSazonov commented:
Based on this I suggest to call `Format-TimeSpanHuman` from DotNetTypes_format_ps1xml.cs and place the cmdlet in Microsoft.PowerShell.Utility.psm1 (with your code).
Later it will be easier to localize it in-place or use a third-party package.

rkeithhill commented:
You can do localization in a format file e.g. (from FileSystem.format.ps1xml):
```xml
<CustomItem>
    <Text AssemblyName="System.Management.Automation" BaseName="FileSystemProviderStrings" ResourceId="DirectoryDisplayGrouping"/>
    <ExpressionBinding>
        <ScriptBlock>
            $_.PSParentPath.Replace("Microsoft.PowerShell.Core\FileSystem::", "")                                                  
        </ScriptBlock>
    </ExpressionBinding>
    <NewLine/>
</CustomItem> 
```
Seems like you just need various localized strings slipped into SMA e.g.: `{0} days,`, `{0} hours,` ... `{0}.{1} seconds` and the bit for formatting script that @lzybkr provided.  This would be a *sweet* enhancement.

iSazonov commented:
@rkeithhill The problem is that different languages require not only different strings but different code.

lzybkr commented:
I tried a small experiment with Humanizer to see all of the strings it produces for a few different timespans and every culture it supports, see the results below.  This experiment also found a bug in Humanizer which I reported.

My conclusion - it might be nice (but maybe not required) to special case a few things like "no time" and to use the equivalent of "an hour" instead of "1 hour" for some languages, but it's probably not much code.

BTW, it's not that hard to create a custom `ToString` method using ETS that calls into C#, here is one [example](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/src/System.Management.Automation/engine/Types_Ps1Xml.generated.cs#L39).

```
#19 PS> [timespan]::new(1,2,3,4,5) | % { $t = $_; $cultures | % { "Culture: {0,-10} {1}" -f $_,[Humanizer.TimeSpanHumanizeExtensions]::Humanize($t, 5, [CultureInfo]$_) } }
Culture: af         1 dag, 2 ure, 3 minute, 4 sekondes, 5 millisekondes
Culture: ar         يوم واحد, ساعتين, 3 دقائق, 4 ثوان, 5 أجزاء من الثانية
Culture: bg         един ден, 2 часа, 3 минути, 4 секунди, 5 милисекунди
Culture: bn-BD      এক দিন, 2 ঘণ্টা, 3 মিনিট, 4 সেকেন্ড, 5 মিলিসেকেন্ড
Culture: cs         1 den, 2 hodiny, 3 minuty, 4 sekundy, 5 milisekund
Culture: da         en dag, 2 timer, 3 minutter, 4 sekunder, 5 millisekunder
Culture: de         Ein Tag, 2 Stunden, 3 Minuten, 4 Sekunden, 5 Millisekunden
Culture: el         1 day, 2 hours, 3 minutes, 4 seconds, 5 milliseconds
Culture: es         un día, 2 horas, 3 minutos, 4 segundos, 5 milisegundos
Culture: fa         یک روز, 2 ساعت, 3 دقیقه, 4 ثانیه, 5 میلی ثانیه
Culture: fi-FI      1 day, 2 hours, 3 minutes, 4 seconds, 5 milliseconds
Culture: fr         1 jour, 2 heures, 3 minutes, 4 secondes, 5 millisecondes
Culture: fr-BE      1 jour, 2 heures, 3 minutes, 4 secondes, 5 millisecondes
Culture: he         יום, שעתיים, 3 דקות, 4 שניות, 5 אלפיות שנייה
Culture: hr         1 dan, 2 sata, 3 minute, 4 sekunde, 5 milisekundi
Culture: hu         1 nap, 2 óra, 3 perc, 4 másodperc, 5 ezredmásodperc
Culture: id         1 hari, 2 jam, 3 menit, 4 detik, 5 milidetik
Culture: it         1 giorno, 2 ore, 3 minuti, 4 secondi, 5 millisecondi
Culture: ja         1 日, 2 時間, 3 分, 4 秒, 5 ミリ秒
Culture: nb         en dag, 2 timer, 3 minutter, 4 sekunder, 5 millisekunder
Culture: nb-NO      en dag, 2 timer, 3 minutter, 4 sekunder, 5 millisekunder
Culture: nl         1 dag, 2 uur, 3 minuten, 4 seconden, 5 milliseconden
Culture: pl         1 dzień, 2 godziny, 3 minuty, 4 sekundy, 5 milisekund
Culture: pt         1 day, 2 hours, 3 minutes, 4 seconds, 5 milliseconds
Culture: ro         1 zi, 2 ore, 3 minute, 4 secunde, 5 milisecunde
Culture: ru         один день, 2 часа, 3 минуты, 4 секунды, 5 миллисекунд
Culture: sk         1 deň, 2 hodiny, 3 minúty, 4 sekundy, 5 milisekúnd
Culture: sl         1 dan, 2 uri, 3 minute, 4 sekunde, 5 milisekund
Culture: sr         1 дан, 2 сата, 3 минута, 4 секунде, 5 милисекунди
Culture: sr-Latn    1 dan, 2 sata, 3 minuta, 4 sekunde, 5 milisekundi
Culture: sv         1 dag, 2 timmar, 3 minuter, 4 sekunder, 5 millisekunder
Culture: tr         1 gün, 2 saat, 3 dakika, 4 saniye, 5 milisaniye
Culture: uk         один день, 2 години, 3 хвилини, 4 секунди, 5 мілісекунд
Culture: uz-Cyrl-UZ 1 кун, 2 соат, 3 минут, 4 секунд, 5 миллисекунд
Culture: uz-Latn-UZ 1 kun, 2 soat, 3 minut, 4 sekund, 5 millisekund
Culture: vi         1 ngày, 2 giờ, 3 phút, 4 giây, 5 phần ngàn giây
Culture: zh-CN      1 天, 2 小时, 3 分, 4 秒, 5 毫秒
Culture: zh-Hans    1 天, 2 小时, 3 分, 4 秒, 5 毫秒
Culture: zh-Hant    1 天, 2 小時, 3 分, 4 秒, 5 毫秒

#20 PS> [timespan]::new(0,0,0,0,0) | % { $t = $_; $cultures | % { "Culture: {0,-10} {1}" -f $_,[Humanizer.TimeSpanHumanizeExtensions]::Humanize($t, 5, [CultureInfo]$_) } }
Culture: af         geen tyd
Culture: ar         حالاً
Culture: bg         няма време
Culture: bn-BD      শূন্য সময়
Culture: cs         není čas
Culture: da         ingen tid
Culture: de         Keine Zeit
Culture: el         no time
Culture: es         0 segundos
Culture: fa         الآن
Culture: fi-FI      no time
Culture: fr         pas de temps
Culture: fr-BE      pas de temps
Culture: he         אין זמן
Culture: hr         bez podatka o vremenu
Culture: hu         nincs idő
Culture: id         waktu kosong
Culture: it         0 secondi
Culture: ja         0 秒
Culture: nb         ingen tid
Culture: nb-NO      ingen tid
Culture: nl         geen tijd
Culture: pl         brak czasu
Culture: pt         no time
Culture: ro         niciun timp
Culture: ru         нет времени
Culture: sk         žiadny čas
Culture: sl         nič časa
Culture: sr         без протеклог времена
Culture: sr-Latn    bez proteklog vremena
Culture: sv         ingen tid
Culture: tr         zaman farkı yok
Culture: uk         без часу
Culture: uz-Cyrl-UZ вақт йўқ
Culture: uz-Latn-UZ vaqt yo`q
Culture: vi         không giờ
Culture: zh-CN      没有时间
Culture: zh-Hans    没有时间
Culture: zh-Hant    沒有時間

#21 PS> [timespan]::new(1,1,1,1,1) | % { $t = $_; $cultures | % { "Culture: {0,-10} {1}" -f $_,[Humanizer.TimeSpanHumanizeExtensions]::Humanize($t, 5, [CultureInfo]$_) } }
Culture: af         1 dag, 1 uur, 1 minuut, 1 sekond, 1 millisekond
Culture: ar         يوم واحد, ساعة واحدة, دقيقة واحدة, ثانية واحدة, جزء من الثانية
Culture: bg         един ден, един час, една минута, една секунда, една милисекунда
Culture: bn-BD      এক দিন, এক ঘণ্টা, এক মিনিট, এক সেকেন্ড, এক মিলিসেকেন্ড
Culture: cs         1 den, 1 hodina, 1 minuta, 1 sekunda, 1 milisekunda
Culture: da         en dag, en time, et minut, et sekund, et millisekund
Culture: de         Ein Tag, Eine Stunde, Eine Minute, Eine Sekunde, Eine Millisekunde
Culture: el         1 day, 1 hour, 1 minute, 1 second, 1 millisecond
Culture: es         un día, una hora, un minuto, un segundo, un milisegundo
Culture: fa         یک روز, یک ساعت, یک دقیقه, یک ثانیه, یک میلی ثانیه
Culture: fi-FI      1 day, 1 hour, 1 minute, 1 second, 1 millisecond
Culture: fr         1 jour, 1 heure, 1 minute, 1 seconde, 1 milliseconde
Culture: fr-BE      1 jour, 1 heure, 1 minute, 1 seconde, 1 milliseconde
Culture: he         יום, שעה, דקה, שנייה, אלפית שנייה
Culture: hr         1 dan, 1 sat, 1 minuta, 1 sekunda, 1 milisekunda
Culture: hu         1 nap, 1 óra, 1 perc, 1 másodperc, 1 ezredmásodperc
Culture: id         1 hari, 1 jam, 1 menit, 1 detik, 1 milidetik
Culture: it         1 giorno, 1 ora, 1 minuto, 1 secondo, 1 millisecondo
Culture: ja         1 日, 1 時間, 1 分, 1 秒, 1 ミリ秒
Culture: nb         en dag, en time, ett minutt, ett sekund, ett millisekund
Culture: nb-NO      en dag, en time, ett minutt, ett sekund, ett millisekund
Culture: nl         1 dag, 1 uur, 1 minuut, 1 seconde, 1 milliseconde
Culture: pl         1 dzień, 1 godzina, 1 minuta, 1 sekunda, 1 milisekunda
Culture: pt         1 day, 1 hour, 1 minute, 1 second, 1 millisecond
Culture: ro         1 zi, 1 oră, 1 minut, 1 secundă, 1 milisecundă
Culture: ru         один день, один час, одна минута, одна секунда, одна миллисекунда
Culture: sk         1 deň, 1 hodina, 1 minúta, 1 sekunda, 1 milisekunda
Culture: sl         1 dan, 1 ura, 1 minuta, 1 sekunda, 1 milisekunda
Culture: sr         1 дан, 1 сат, 1 минут, 1 секунда, 1 милисекунда
Culture: sr-Latn    1 dan, 1 sat, 1 minut, 1 sekunda, 1 milisekunda
Culture: sv         1 dag, 1 timma, 1 minut, 1 sekund, 1 millisekund
Culture: tr         1 gün, 1 saat, 1 dakika, 1 saniye, 1 milisaniye
Culture: uk         один день, одна година, одна хвилина, одна секунда, одна мілісекунда
Culture: uz-Cyrl-UZ 1 кун, 1 соат, 1 минут, 1 секунд, 1 миллисекунд
Culture: uz-Latn-UZ 1 kun, 1 soat, 1 minut, 1 sekund, 1 millisekund
Culture: vi         1 ngày, 1 giờ, 1 phút, 1 giây, 1 phần ngàn giây
Culture: zh-CN      1 天, 1 小时, 1 分, 1 秒, 1 毫秒
Culture: zh-Hans    1 天, 1 小时, 1 分, 1 秒, 1 毫秒
Culture: zh-Hant    1 天, 1 小時, 1 分, 1 秒, 1 毫秒
```

iSazonov commented:
Is Powershell the right place to do it? Perhaps it is better to do it in .Net?
Whether to override `ToString`? Perhaps better to create `ToHumanString` or  IFormatProvider?

lzybkr commented:
It'd be great if .Net added a nice api, we'd use it for sure. But if they don't, we should do something.

We should only override `ToString` when the existing one is completely useless. So yeah, we can just add some static method, call it whatever you want, and call that from the formatting entry.

rkeithhill commented:
You could also create a static extension method for  `ToString` on TimeSpan objects with a `humanize` boolean parameter.

iSazonov commented:
@rkeithhill Thanks! "boolean" does not tell us about "human". 😕  Could be done IFormatProvider but a static `ToHumanString` method is more user friendly.

@lzybkr What is the correct place for this "humanizer" code?

rkeithhill commented:
@iSazonov it can "tell you" *if* you use C#'s named parameter functionality e.g. `timespan.ToString(humanize: true);` but it does require discipline by the caller.

iSazonov commented:
@rkeithhill It's interesting! Is it possible to make such a call from Powershell?

lzybkr commented:
Named parameters to .Net methods are not yet supported, see https://github.com/PowerShell/PowerShell-RFC/issues/6

rkeithhill commented:
Well that and PowerShell doesn't have language support for static extension methods.  However, you can invoke static extension methods the old fashioned way - like any other static method.  :-)

iSazonov commented:
@lzybkr What is the correct place for this "humanizer" code and resource file?

lzybkr commented:
Good question. We have a few similar methods, so maybe [here](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/src/System.Management.Automation/engine/CodeMethods.cs#L19) or [here](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/src/System.Management.Automation/engine/MshObject.cs#L2412), or maybe that partial class should just be put in a new file.

iSazonov commented:
@lzybkr Thanks for the idea. I don't lose hope that it will be done in .Net so I suppose that a new file in [engine](https://github.com/PowerShell/PowerShell/tree/master/src/System.Management.Automation/engine) folder will be better.

iSazonov committed:
Add ToHumanString

1. Add code method ToHumanString
2. Add resorce file
3. Add formating for TimeSpan
4. Add default HumanFormatter class
5. Add Humanizer tests
6. Add test for default output to console

iSazonov commented:
I took the next step.
There is problems.
1. I have not been able to compile the resource file named `Humanizer.en-Us.resx`
2. Generated resource class has no `GetString(String, CultureInfo)` . Had to make a call `Humanizer.Culture = Culture;` Is it acceptable?
3. The algorithm is not made common to all languages. Should we do it now?
4. How test this? As timespan formatting for en-Us only or/and full tests of ToHumanString for all cultures?

iSazonov committed:
Fix Travis test

iSazonov commented:
Reopen to restart CI.

lzybkr commented on src/System.Management.Automation/resources/Humanizer.resx:
```diff
@@ -0,0 +1,165 @@
+    <value>{0} days</value> 
+  </data> 
+  <data name="TimeSpanHumanize_MultipleDays_Paucal" xml:space="preserve"> 
+    <value>{0} days</value> 
```
I'm not sure we want Paucal numbers. The lack of precision seems wrong in platform tool for programming. For example, it may be linguistically correct to say `a few seconds`, that lacks the precision people would expect from PowerShell.

iSazonov commented on src/System.Management.Automation/engine/Humanizer.cs:
```diff
@@ -0,0 +1,132 @@
+        /// </summary>
+        /// <param name="time">time is TimeSpan object</param>
+
+        public static string ToHumanString(TimeSpan time)
```
The easiest way is to simply move here all the code and resources of the `Humanizer` project. If it is a valid solution then let's do it.

lzybkr replied:
Based on my experiments, I prefer my implementation over humanizer for things like `Measure-Command`, so my preference is to not adopt their code.

iSazonov replied:
Thanks for the conclusion.
I'm not a polyglot. So I needed some landmark like this `Humanizer` project. Now it is clear that each language requires _up to 4 resource strings_ and maybe their _code_. And I'm ready to forget about `Humanizer`.

I'm willing to rename things but don't feel free with English terms. I need your help.

Above I mentioned that I have not been able to compile the resource file named  Humanizer.en-Us.resx  Please comment this. Can we use existing way of working with resources here (or directly use a resource manager)? If Yes, then it requires improvements.
